### PR TITLE
feat: add example configurations and end-to-end tests

### PR DESCRIPTION
**Pull request by vicoooo26**
## Summary
- 11 example YAML files: MPI (simple, with-launcher), PyTorch (simple, standalone, tensorboard-mounts, with-master), TensorFlow (simple, with-roles), test fixtures (configmap, features, secret)
- E2E tests: suite setup, MPI/PyTorch/TFJob submission, dry-run, format, check, CLI overrides, YAML submit/delete
- Integration tests: flow, loader, multijob, override, output, stop/resume, validation, error display

Part of the Arena v2 CLI series (PR 7/7). Depends on #1465, #1466, #1467, #1468, #1469, #1470.

**Note:** Review only the last commit for this PR's changes.

## Test plan
- [ ] `go test ./test/... -v` passes (integration tests)
- [ ] `go vet ./test/...` passes
- [ ] E2E tests compile: `go test -tags v2e2e ./test/e2e/... -v -run None`
- [ ] Example YAMLs are valid: `go run ./cmd/arena-v2/ submit --dry-run examples/v2/pytorch-simple.yaml`

### CODE REVIEW

**Comment by cheyang** [comment]
@vicoooo26 — this PR is blocked from merging by a **failing DCO check**.
The DCO bot requires every commit to carry a `Signed-off-by:` trailer; without it the PR cannot land regardless of `lgtm` / `approved` labels or otherwise-green CI.

Please resolve this now. For a single-commit PR:

```
git commit --amend -s
git push --force-with-lease
```

For multiple commits on this branch:

```
git rebase --signoff origin/develop-v2
git push --force-with-lease
```

The code review side is finished and any merge labels already on the PR will take effect once DCO turns green. Reference: https://github.com/apps/dco

**Comment by cheyang** [comment]
@gemini-code-assist /review

**Comment by cheyang** [comment]
Thanks for adding the v2 examples and e2e coverage. Given the size of the PR I focused on high-risk areas. The main blocker is `examples/v2/test-secret.yaml`, which demonstrates printing a private SSH key to stdout — that's a security anti-pattern for a user-facing example. There is also a failing DCO check that will need the commits signed off before this can merge. Beyond that, I left a few notes on e2e test depth and example usability.

File: test/testdata/secret.yaml [hunk 1]
```diff
@@ -0,0 +1,19 @@
+version: 0.1.0
+name: test-secret
+framework:
+  name: pytorch
+image: pytorch/pytorch:2.1
+run: cat /root/.ssh/id_rsa
```
Comment (cheyang) [inline test/testdata/secret.yaml]: The example runs `cat /root/.ssh/id_rsa`, which prints a private SSH key to stdout/logs. This is a dangerous pattern to ship as a user-facing example: anyone who copies it can leak a secret into container logs. Please change the command to something benign (e.g., `cat /etc/hostname` or a no-op) so the example demonstrates the secret mount without exposing the key.

File: test/e2e/mpi_test.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+		By("Cleaning up")
+		delCmd := exec.Command(arenaV2Bin, "job", "delete", jobName,
+			"--namespace", namespace)
+		delCmd.Stdout = &out
+		delCmd.Stderr = &out
+		_ = delCmd.Run()
+	}
+
+	It("MPI job lifecycle", func() {
+		frameworkLifecycle("mpi", "mpi:latest")
```
Comment (cheyang) [inline test/e2e/mpi_test.go]: The e2e lifecycle tests use placeholder images like `mpi:latest`, `horovod:latest`, and `deepspeed:latest` and only verify that the CR is created, listed, and deleted. They don't wait for pods to become ready or jobs to complete, so the tests can pass even when the images can't be pulled or the workload never runs. Consider adding a readiness wait or clearly documenting that these are CRUD smoke tests.

File: test/e2e/suite_test.go [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+		arenaV2Bin = p
+	} else {
+		Fail("arena-v2 binary not found in bin/ or PATH — run `make arena-v2` first")
+	}
+})
+
+func createTempYAML(content string) (string, error) {
+	tmpDir := os.TempDir()
+	path := filepath.Join(tmpDir, fmt.Sprintf("arena-test-%d.yaml", GinkgoRandomSeed()))
+	err := os.WriteFile(path, []byte(content), 0644)
```
Comment (cheyang) [inline test/e2e/suite_test.go]: `createTempYAML` writes test files to `/tmp` with mode `0644` and a name derived from `GinkgoRandomSeed()`. That's world-readable and somewhat predictable. If these temp files ever carry secrets this would leak them; using `0600` and a random suffix would be safer.

File: examples/v2/mpi-simple.yaml [hunk 4]
```diff
@@ -0,0 +1,15 @@
+version: 0.1.0
+name: mpi-example
+image: openmpi:4.1
```
Comment (cheyang) [inline examples/v2/mpi-simple.yaml]: Several examples use unqualified image names (`openmpi:4.1`, `pytorch:2.1`, `tensorflow:2.15`) without a registry. Users copying these directly will likely hit `ImagePullBackOff` unless the images already exist locally. Consider using fully-qualified public images or adding a comment telling users to replace the image.

File: test/testdata/configmap.yaml [hunk 5]
```diff
@@ -0,0 +1,19 @@
+version: 0.1.0
+name: test-configmap
```
Comment (cheyang) [inline test/testdata/configmap.yaml:2]: The `test-configmap.yaml` and `test-secret.yaml` files are named like test fixtures and aren't referenced by any test. If they're meant as user examples, consider giving them clearer names; if they're fixtures, move them under a `testdata/` directory so users don't copy them as canonical examples.

**Comment by cheyang** [comment]
@vicoooo26 — this PR is blocked from merging by a **failing DCO check**.
The DCO bot requires every commit to carry a `Signed-off-by:` trailer; without it the PR cannot land regardless of `lgtm` / `approved` labels or otherwise-green CI.

Please resolve this now. For a single-commit PR:

```
git commit --amend -s
git push --force-with-lease
```

For multiple commits on this branch:

```
git rebase --signoff origin/develop-v2
git push --force-with-lease
```

The code review side is finished and any merge labels already on the PR will take effect once DCO turns green. Reference: https://github.com/apps/dco

**Comment by cheyang** [comment]
@gemini-code-assist /review

File: go.mod [hunk 6]
```diff
@@ -1,3 +1,63 @@
...
+	k8s.io/api v0.36.2
+	k8s.io/apimachinery v0.36.2
+	k8s.io/client-go v0.36.2
+	k8s.io/klog/v2 v2.140.0
+)
+
+require (
+	github.com/Masterminds/semver/v3 v3.5.0 // indirect
+	github.com/davecgh/go-spew v1.1.2-0.20180830191138-d8f796af33cc // indirect
+	github.com/emicklei/go-restful/v3 v3.13.0 // indirect
```
Comment (cheyang) [inline go.mod:19]: Adding `helm.sh/helm/v3` as a direct dependency conflicts with the project constraint in `AGENTS.md` (#2): v2 must not depend on Helm. Pulling in all of Helm just for `--set` parsing significantly expands the dependency surface and contradicts the design goal of eliminating Helm from Arena v2. Please implement a small standalone `--set` parser or use a minimal key-value library instead.

File: pkg/task/setter.go [hunk 7]
```diff
@@ -0,0 +1,154 @@
+package task
+
+import (
+	"fmt"
+	"strings"
+
+	"gopkg.in/yaml.v3"
+	"helm.sh/helm/v3/pkg/strvals"
```
Comment (cheyang) [inline pkg/task/setter.go]: Importing `helm.sh/helm/v3/pkg/strvals` ties v2 code to Helm. Since the project explicitly forbids Helm dependencies in v2, this import needs to be replaced with a local implementation of dotted key-value parsing before the PR can land.

File: examples/v2/mpi-simple.yaml (hunk 4 above)
Comment (cheyang) [inline examples/v2/mpi-simple.yaml]: The image `openmpi:4.1` lacks a registry/namespace, so Kubernetes will resolve it against the default registry (usually Docker Hub). This makes the example fragile and environment-dependent. Other examples have the same issue (`examples/v2/mpi-with-launcher.yaml`, `examples/v2/pytorch-with-master.yaml` with `pytorch:2.1`, `examples/v2/tf-with-roles.yaml` with `tensorflow:2.15`). Please qualify all example images with a full registry path and a stable tag.

File: test/testdata/secret.yaml [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+worker:
```
Comment (cheyang) [inline test/testdata/secret.yaml:7]: `run: cat /root/.ssh/id_rsa` demonstrates reading a private SSH key back to stdout. Even though the key in `prereq-secret.yaml` is fake, this is still a security anti-pattern for a user-facing test fixture. Please change the command to something that does not expose sensitive key material, e.g. `ls /root/.ssh/` or `cat /secrets/username`.

File: test/e2e/mpi_test.go (hunk 2 above)
Comment (cheyang) [inline test/e2e/mpi_test.go]: The MPI, Horovod, and DeepSpeed lifecycle tests only assert that `job get` output contains the framework name. They do not validate the generated CRD structure, resources, or that the correct MPIJob apiVersion was used. Consider strengthening these to inspect the submitted CR with `kubectl get` + JSON assertions, similar to `storage_test.go`.

**Comment by cheyang** [comment]
@vicoooo26 — this PR is blocked from merging by a **failing DCO check**.
The DCO bot requires every commit to carry a `Signed-off-by:` trailer; without it the PR cannot land regardless of `lgtm` / `approved` labels or otherwise-green CI.

Please resolve this now. For a single-commit PR:

```
git commit --amend -s
git push --force-with-lease
```

For multiple commits on this branch:

```
git rebase --signoff origin/develop-v2
git push --force-with-lease
```

The code review side is finished and any merge labels already on the PR will take effect once DCO turns green. Reference: https://github.com/apps/dco

**Comment by cheyang** [comment]
Thanks for the new examples, e2e suite, and lifecycle commands. This round still has the same blocking issues as before: the DCO check is failing because two commits have a sign-off email that does not match their author email, `go.mod` still depends on `helm.sh/helm/v3`, several examples use unqualified image names, and `test/testdata/secret.yaml` still demonstrates reading a private SSH key. I also flagged that gang scheduling currently sets `minAvailable=0`, which disables gang semantics. Please address the blocking issues; the non-blocking risks (partial-submission resource leaks and cleanup error swallowing) can be handled in a follow-up if you prefer.

File: go.mod [hunk 9]
```diff
@@ -1,3 +1,63 @@
...
-go 1.25.1
+go 1.26.1
+
+require (
+	github.com/onsi/ginkgo/v2 v2.32.0
+	github.com/onsi/gomega v1.42.1
+	github.com/spf13/cobra v1.10.2
+	github.com/stretchr/testify v1.11.1
+	gopkg.in/yaml.v3 v3.0.1
+	helm.sh/helm/v3 v3.21.3
```
Comment (cheyang) [inline go.mod]: Adding `helm.sh/helm/v3` pulls in a large, Kubernetes-release-tied dependency tree just to use `pkg/strvals` for `--set` parsing. This conflicts with the v2 project's goal of a lightweight CLI built on client-go. Please vendor or reimplement the small `--set` parser locally and remove the Helm v3 module so `go.mod` stays minimal.

File: examples/v2/mpi-simple.yaml (hunk 4 above)
Comment (cheyang) [inline examples/v2/mpi-simple.yaml]: Several example YAMLs use short image names such as `openmpi:4.1`, `pytorch:2.1`, and `tensorflow:2.15` without a registry. These are ambiguous, can resolve differently on different hosts, and may pull from untrusted defaults. Please use fully-qualified image references (e.g., `docker.io/library/...` or a documented project registry) in all examples and in the e2e tests that reuse them.

File: test/testdata/secret.yaml (hunk 1 above)
Comment (cheyang) [inline test/testdata/secret.yaml]: This fixture demonstrates `run: cat /root/.ssh/id_rsa`, which reads a private SSH key into logs. Even as testdata, this is a security anti-pattern and could be copied into real jobs. Please replace the command with a non-sensitive validation step (e.g., `test -f /root/.ssh/id_rsa && echo ok`) and avoid exposing private key material.

File: pkg/provider/interface.go [hunk 10]
```diff
@@ -0,0 +1,1060 @@
...
+	}
+	sp := map[string]interface{}{}
+	if s.Queue != "" {
+		sp["queue"] = s.Queue
+	}
+	if s.PriorityClassName != "" {
+		sp["priorityClass"] = s.PriorityClassName
+	}
+	if s.Gang.Enabled {
+		sp["minAvailable"] = int64(0)
```
Comment (cheyang) [inline pkg/provider/interface.go]: When `scheduling.gang.enabled` is true, `buildSchedulingPolicy` hard-codes `minAvailable` to `0`. That disables gang semantics entirely because the scheduler sees no minimum pods. It should be set to the total expected replicas (or made configurable), not zero.

File: pkg/cli/lifecycle.go [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+	"github.com/kubeflow/arena/pkg/client"
+	"github.com/kubeflow/arena/pkg/constants"
+	"github.com/kubeflow/arena/pkg/provider"
+	"github.com/kubeflow/arena/pkg/task"
+)
+
+// createJobResources creates auxiliary resources after the main CRD is created.
+// It creates a ConfigMap anchor with ownerReferences pointing to the CRD,
+// and optionally creates TensorBoard Deployment and Service resources.
+func createJobResources(ctx context.Context, crd *unstructured.Unstructured, t *task.Task, k8sClient *client.Client, p provider.Provider) error {
```
Comment (cheyang) [inline pkg/cli/lifecycle.go:43]: `createJobResources` creates the ConfigMap, RBAC resources, and TensorBoard resources sequentially. If the TensorBoard Service or RBAC creation fails, the previously created ConfigMap/RBAC objects are left behind. Consider either creating all resources in a transaction-like loop with cleanup, or applying ownerReferences consistently so the main CRD's deletion cascades everything (ownerRefs are present, but leaks still occur while the job exists).

**Comment by cheyang** [comment]
@vicoooo26 — this PR is blocked from merging by a **failing DCO check**.
The DCO bot requires every commit to carry a `Signed-off-by:` trailer; without it the PR cannot land regardless of `lgtm` / `approved` labels or otherwise-green CI.

Please resolve this now. For a single-commit PR:

```
git commit --amend -s
git push --force-with-lease
```

For multiple commits on this branch:

```
git rebase --signoff origin/develop-v2
git push --force-with-lease
```

The code review side is finished and any merge labels already on the PR will take effect once DCO turns green. Reference: https://github.com/apps/dco

**Comment by cheyang** [comment]
@gemini-code-assist /review

**Comment by cheyang** [comment]
Thanks for the updates. I see that `test-secret.yaml` was removed, the `secret.yaml` example no longer echoes a private SSH key, and `examples/v2/` images are now registry-qualified. Gang scheduling also correctly defaults `minAvailable` to at least 1. However, a few blockers from the previous round are still present and need to be addressed before this can merge:

1. `helm.sh/helm/v3` is still a direct dependency (used in `pkg/task/setter.go`).
2. Default images in `pkg/constants/constants.go` and fixtures in `test/testdata/` are still unqualified.
3. The DCO check is failing (`ACTION_REQUIRED`) because the sign-off email does not match the author email on some commits.

I also noted that the E2E smoke tests use unqualified placeholder images; qualifying those would keep the test suite consistent with the examples. Once the blockers above are resolved, this should be ready for another look.

File: pkg/task/setter.go (hunk 7 above)
Comment (cheyang) [inline pkg/task/setter.go]: This import keeps `helm.sh/helm/v3` in `go.mod`. The previous round flagged the Helm v3 dependency as violating project constraints, and it is still present. Please vendor or re-implement the small `strvals.ParseInto` behaviour you need instead of depending on the full Helm chart/CLI module.

File: pkg/constants/constants.go [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+
+// Default images for sync init containers.
+const (
+	// DefaultGitSyncImage is the git-sync sidecar used to clone a Git repository
+	// into the training pod before the main container starts.
+	DefaultGitSyncImage = "registry.k8s.io/git-sync/git-sync:v3.3.5"
+
+	// DefaultRsyncImage is the rsync image used to copy a remote directory
+	// into the training pod via rsync.
+	DefaultRsyncImage = "instrumentisto/rsync-ssh:3.18"
```
Comment (cheyang) [inline pkg/constants/constants.go:13]: `DefaultRsyncImage`, `DefaultHDFSImage`, and `DefaultTensorBoardImage` are not registry-qualified. This makes them ambiguous when clusters use private mirrors, custom default registries, or image policies. Please prefix them with `docker.io/` (or the correct upstream registry) so the default behaviour is explicit and reproducible.

File: test/testdata/configmap.yaml [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+framework:
+  name: pytorch
+image: pytorch/pytorch:2.1
```
Comment (cheyang) [inline test/testdata/configmap.yaml]: The `test/testdata/` fixtures (`configmap.yaml`, `scheduling.yaml`, `secret.yaml`) still use unqualified images (`pytorch/pytorch:2.1`, `pytorch:2.1`, `busybox`). These files are user-facing examples in tests and should be consistent with `examples/v2/`, which now correctly uses `docker.io/...`. Please qualify them.

File: go.mod (hunk 9 above)
Comment (cheyang) [inline go.mod]: The DCO status check is currently `ACTION_REQUIRED`. Although the commit messages contain `Signed-off-by:` trailers, the sign-off email (`vico24826@gmail.com`) does not match the author email on several commits (`30412827+vicoooo26@users.noreply.github.com`). Please amend those commits with a matching sign-off (`git commit --amend -s` or `git rebase --signoff`) and force-push.

File: test/e2e/cli_overrides_test.go [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+		delCmd.Stderr = &out
+		_ = delCmd.Run()
+	})
+
+	It("should show detailed job information", func() {
+		var out bytes.Buffer
+		submitCmd := exec.Command(arenaV2Bin, "job", "submit", "pytorch",
+			"--name", jobName,
+			"--namespace", namespace,
+			"--image", "pytorch:2.1",
```
Comment (cheyang) [inline test/e2e/cli_overrides_test.go]: Several E2E tests pass unqualified placeholder images (`pytorch:2.1`, `horovod:latest`, `openmpi:4.1`, etc.) to the CLI. Because these tests do not wait for pods, the images are not actually pulled, but they still set a bad example and could break if the test suite is later extended to wait for readiness. Consider qualifying them with `docker.io/` for consistency.
